File: packages/python/sheet_compressor/encodings/anchor.py
"""Structural-anchor skeleton encoding (SPEC §3)."""

from typing import Any, Mapping

from ..address import a1
from .escape import escape_value


def encode_anchor(
    grid: Mapping[str, Any],
    detection: Mapping[str, set],
    token_counter,
) -> dict:
    rows = grid.get("rows", [])
    origin = grid["origin"]
    kept_rows = detection["kept_rows"]
    kept_cols = detection["kept_cols"]

    cells = []
    lines = []
    for r in range(len(rows)):
        if r not in kept_rows:
            continue
        row = rows[r]
        tokens = []
        for c in range(len(row)):
            if c not in kept_cols:
                continue
            value = row[c]
            # SPEC §3.1: only literal "" is empty.
            if value == "":
                continue
            address = a1(origin["row"] + r, origin["col"] + c)
            cells.append({"address": address, "value": value})
            tokens.append(f"{address},{escape_value(value)}")
        if tokens:
            lines.append("|".join(tokens))

    string = "\n".join(lines)
    json_obj = {
        "encoding": "anchor-skeleton",
        "version": 0,
        "origin": {"row": origin["row"], "col": origin["col"]},
        "cells": cells,
    }
    return {"string": string, "json": json_obj, "tokenEstimate": token_counter(string)}

File: packages/go/sheetcompressor/anchor.go
package sheetcompressor

import "strings"

func encodeAnchor(
	grid *Grid,
	detection AnchorDetection,
	tokenCounter TokenCounter,
) Encoding[AnchorJSON] {
	// Pre-size to avoid the common "tiny grid → many tiny appends" allocations.
	cells := make([]AnchorCell, 0)
	var lines []string

	for r, row := range grid.Rows {
		if _, ok := detection.KeptRows[r]; !ok {
			continue
		}
		var tokens []string
		for c := 0; c < len(row); c++ {
			if _, ok := detection.KeptCols[c]; !ok {
				continue
			}
			value := row[c]
			// SPEC §3.1: only literal "" is empty.
			if value == "" {
				continue
			}
			address := a1(grid.Origin.Row+r, grid.Origin.Col+c)
			cells = append(cells, AnchorCell{Address: address, Value: value})
			tokens = append(tokens, address+","+escapeValue(value))
		}
		// SPEC §3.2: fully-empty rows are dropped (no blank line emitted).
		if len(tokens) > 0 {
			lines = append(lines, strings.Join(tokens, "|"))
		}
	}

	s := strings.Join(lines, "\n")
	j := AnchorJSON{
		Encoding: "anchor-skeleton",
		Version:  0,
		Origin:   grid.Origin,
		Cells:    cells,
	}
	return Encoding[AnchorJSON]{String: s, JSON: j, TokenEstimate: tokenCounter(s)}
}
